### PR TITLE
chore: Introduce `WorkflowPhase`

### PR DESCRIPTION
**Pull request by alexec**
Signed-off-by: Alex Collins <alex_collins@intuit.com>

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: pkg/apis/workflow/v1alpha1/workflow_phase.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+const (
+	WorkflowUnknown   WorkflowPhase = ""
+	WorkflowPending   WorkflowPhase = "Pending" // pending some set-up - rarely used
+	WorkflowRunning   WorkflowPhase = "Running" // any node has started; pods might not be running yet, the workflow maybe suspended too
+	WorkflowSucceeded WorkflowPhase = "Succeeded"
+	WorkflowFailed    WorkflowPhase = "Failed" // it maybe that the the workflow was terminated
+	WorkflowError     WorkflowPhase = "Error"
+)
+
+func (p WorkflowPhase) Completed() bool {
```
Comment (simster7) [inline pkg/apis/workflow/v1alpha1/workflow_phase.go:15]: Should this be fulfilled?

Comment (alexec) [reply pkg/apis/workflow/v1alpha1/workflow_phase.go:15]: I'm not sure either fulfilled or complete should apply to the workflows phase at all. The source of truth is the label. Phase and conditions are just copies of this information.

File: ui/src/models/workflows.ts [hunk 2]
```diff
@@ -848,9 +848,11 @@ export interface MemoizationStatus {
     cacheName: string;
 }
 
-export type NodePhase = '' | 'Pending' | 'Running' | 'Succeeded' | 'Skipped' | 'Failed' | 'Error' | 'Omitted';
+export type WorkflowPhase = 'Pending' | 'Running' | 'Succeeded' | 'Failed' | 'Error';
+
+export const WorkflowPhases: WorkflowPhase[] = ['Pending', 'Running', 'Succeeded', 'Failed', 'Error'];
 
-export const WorkflowPhases: NodePhase[] = ['Pending', 'Running', 'Succeeded', 'Failed', 'Error'];
+export type NodePhase = '' | 'Pending' | 'Running' | 'Succeeded' | 'Skipped' | 'Failed' | 'Error' | 'Omitted';
```
Comment (simster7) [inline ui/src/models/workflows.ts:855]: Nice, update was needed here

File: workflow/controller/operator.go [hunk 3]
```diff
@@ -360,15 +360,19 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
...
-	workflowStatus := node.Phase
+	workflowStatus := map[wfv1.NodePhase]wfv1.WorkflowPhase{
+		wfv1.NodePending:   wfv1.WorkflowPending,
+		wfv1.NodeRunning:   wfv1.WorkflowRunning,
+		wfv1.NodeSucceeded: wfv1.WorkflowSucceeded,
+		wfv1.NodeSkipped:   wfv1.WorkflowSucceeded,
+		wfv1.NodeFailed:    wfv1.WorkflowFailed,
+		wfv1.NodeError:     wfv1.WorkflowError,
+		wfv1.NodeOmitted:   wfv1.WorkflowSucceeded,
+	}[node.Phase]
```
Comment (simster7) [inline workflow/controller/operator.go:371]: Perhaps extract this to a function? Could potentially be useful elsewhere

**Review by simster7 — APPROVED** [review]

File: workflow/controller/operator.go (hunk 3 above)
Comment (alexec) [reply workflow/controller/operator.go:371]: I will do that - I can add a test then

**Pull request merged**